Move OpenTelemetry integration to use Instrumenter

radcortez opened the pull request:
- [x] Replace old manual code with the OTel Instrumenter API: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/using-instrumenter-api.md
- [x] Use JAX-RS Filters to cover invocation calls from `WebTarget` (and MP REST Client) 
- [x] Remove old integration code
- [x] Verify that all tests are ok with the new integration
- [x] Micrometer Vert.x Metrics

radcortez committed:
Move OpenTelemetry integration to use Instrumenter

ebullient commented on extensions/opentelemetry/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java:
```diff
@@ -29,6 +30,7 @@
 import io.quarkus.opentelemetry.runtime.OpenTelemetryRecorder;
 import io.quarkus.opentelemetry.runtime.QuarkusContextStorage;
 import io.quarkus.opentelemetry.runtime.tracing.cdi.WithSpanInterceptor;
+import io.quarkus.opentelemetry.runtime.tracing.restclient.OpenTelemetryClientFilter;
```
watch this import -- depending on what it imports, you may run into trouble with a project w/o this enabled.

Supporting optional deps is why most of the telemetry/metrics support works with strings, not with imported class.getName()

radcortez replied:
This is only used here:

https://github.com/radcortez/quarkus/blob/2b148a3d35cdc9aa52c84d57440a097a36ffd422/extensions/opentelemetry/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java#L113-L118

To add it as an additional class to index if the REST Client is Available, by testing this:

https://github.com/radcortez/quarkus/blob/2b148a3d35cdc9aa52c84d57440a097a36ffd422/extensions/opentelemetry/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java#L48-L58

The `OpenTelemetryClientFilter` is then added via CDI and `@Provider`:

https://github.com/radcortez/quarkus/blob/2b148a3d35cdc9aa52c84d57440a097a36ffd422/extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/restclient/OpenTelemetryClientFilter.java#L35-L36

I think it is fairly safe, but let me add some tests for it.

kenfinnigan commented on extensions/opentelemetry/opentelemetry/runtime/pom.xml:
```diff
@@ -27,6 +27,10 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-vertx</artifactId>
         </dependency>
+        <dependency>
```
This will add a HTTP server to all applications using OpenTelemetry, correct?

Is that desired for Reactive Messaging applications?

radcortez replied:
This was to obtain the Vert.x Routing Context, but I may be able to rewrite this without a direct dependency.

kenfinnigan commented on extensions/opentelemetry/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryBuildItem.java:
```diff
@@ -0,0 +1,17 @@
+import io.quarkus.builder.item.SimpleBuildItem;
+import io.quarkus.runtime.RuntimeValue;
+
+public final class OpenTelemetryBuildItem extends SimpleBuildItem {
```
This is used on a producer side, but not being consumed.

Is it needed?

radcortez replied:
Ah, I guess I added it to use it as a parameter to the Vert.x Tracer, but then I ended up using `GlobalOpenTelemetry.get()`. I'll remove it.

kenfinnigan commented on integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OpenTelemetryTestCase.java:
```diff
@@ -189,7 +189,7 @@ void testEmptyClientPath() {
                 Assertions.assertEquals("/", spanData.get("attr_http.target"));
                 Assertions.assertEquals(pathParamUrl.getAuthority(), spanData.get("attr_http.host"));
                 Assertions.assertEquals("http", spanData.get("attr_http.scheme"));
-                Assertions.assertNull(spanData.get("attr_http.route"));
```
I'm not sure the `route` should be `/` given it represents a path template.

If there is no path, it should probably be null/empty

radcortez replied:
Yes, the spec is not very clear about this. Let me try to clarify this with the OTel group. In the meanwhile, I'll revert this.

kenfinnigan commented on integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/OpenTelemetryTestCase.java:
```diff
@@ -303,7 +303,7 @@ void testSlashClientPath() {
                 Assertions.assertEquals(pathParamUrl.getAuthority(), spanData.get("attr_http.host"));
                 Assertions.assertEquals("http", spanData.get("attr_http.scheme"));
-                Assertions.assertNull(spanData.get("attr_http.route"));
+                Assertions.assertEquals("/", spanData.get("attr_http.target"));
```
This changed from testing `route` to `target`, which is already tested on line 303

radcortez replied:
Correct. Wrong copy paste.

kenfinnigan reviewed (commented):
Was a little difficult to fully grasp given the breadth of changes.

Have you verified the attributes and trace/span shape are identical to existing code and post this change for all situations? I'm unsure of the REST Client situation, as that has large changes. Does the correct trace get created for classic and Reactive with sync and async return types?

Do the traces for Reactive Messaging look good still?

radcortez commented on extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/restclient/OpenTelemetryClientFilter.java:
```diff
@@ -0,0 +1,213 @@
+        if (instrumenter.shouldStart(parentContext, request)) {
+            Context spanContext = instrumenter.start(parentContext, request);
+            Scope scope = spanContext.makeCurrent();
+            request.setProperty("otel.span.client.context", spanContext);
```
These are just local keys to use the JAX-RS Context to pass the `Scope` and the `Context` between the filter request and filter response. We can certainly extract them to static strings. They are not intended to be used outside the Client Filter.

radcortez commented:
> Was a little difficult to fully grasp given the breadth of changes.

Yes, sorry but thanks for reviewing :)

> Have you verified the attributes and trace/span shape are identical to existing code and post this change for all situations? I'm unsure of the REST Client situation, as that has large changes. Does the correct trace get created for classic and Reactive with sync and async return types?
> 
> Do the traces for Reactive Messaging look good still?

Yes, I've added a few more tests in `RestClientOpenTelemetryTest`. Also, the tests in the reactive rest-client pass without any change in `io.quarkus.it.rest.client.BasicTest`. We can probably add a few more.

The filter now also supports the JAX-RS Web Client (it was only MP REST Client before).

radcortez commented:
I've removed the Vert.x Http dependency to obtain the route and used the Vert.x Metrics instead. This is still not correct, because we cannot conflict with the Micrometer Metrics, but it was the only other way to retrieve the route. 

In the meanwhile, I'll discuss with @ebullient on how we can share some things between Vert.x Metrics and Tracing.

radcortez commented:
> I've removed the Vert.x Http dependency to obtain the route and used the Vert.x Metrics instead. This is still not correct, because we cannot conflict with the Micrometer Metrics, but it was the only other way to retrieve the route.
> 
> In the meanwhile, I'll discuss with @ebullient on how we can share some things between Vert.x Metrics and Tracing.

For now, I guess the most pragmatic way is to check if the Micrometer extension is available and only register the OTel metrics portion if is not. I've also modified the Micrometer extension to provide a way to retrieve the selected Vert.x route, so it would work in any case.

ebullient commented on extensions/opentelemetry/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java:
```diff
@@ -43,6 +45,18 @@ public boolean getAsBoolean() {
+        public boolean getAsBoolean() {
+            try {
+                Class.forName("javax.ws.rs.client.ClientRequestFilter", false, Thread.currentThread().getContextClassLoader());
+                return true;
```
See what I do in some of the MicrometerProcessors. You can do this once (static final) so you aren't rechecking every time the boolean supplier is used.

radcortez replied:
Sure.

ebullient commented on extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/vertx/OpenTelemetryHttpServerMetrics.java:
```diff
@@ -0,0 +1,43 @@
+import io.vertx.core.http.impl.HttpServerRequestInternal;
+import io.vertx.core.spi.metrics.HttpServerMetrics;
+import io.vertx.core.spi.observability.HttpRequest;
+
```
Can you add a comment here re: mutual exclusivity with micrometer extension?

radcortez replied:
Sure.

ebullient commented on extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/vertx/OpenTelemetryVertxMetricsFactory.java:
```diff
@@ -0,0 +1,21 @@
+import io.vertx.core.spi.metrics.HttpServerMetrics;
+import io.vertx.core.spi.metrics.VertxMetrics;
+
+public class OpenTelemetryVertxMetricsFactory implements VertxMetricsFactory {
```
Same here. Just to make sure relationship (or presence/absence) of this support w.r.t. micrometer extension is clear

ebullient replied:
I might also consider making OpenTelemetryHttpServerMetrics an inner class so that it is a single unit that fills the gap when micrometer isn't around.

radcortez replied:
Sure.

ebullient commented on integration-tests/opentelemetry-vertx/src/main/resources/reflection-config.json:
```diff
@@ -0,0 +1,74 @@
+[
```
Is this needed because of how we're testing?

radcortez replied:
Yes. This is in the integration test modules, so not a big deal.

ebullient commented on integration-tests/opentelemetry-vertx/src/main/resources/application.properties:
```diff
@@ -0,0 +1 @@
+quarkus.native.additional-build-args =-H:ReflectionConfigurationFiles=reflection-config.json
```
Is this only needed because we're testing/checking span values?

radcortez replied:
Yes. This is in the integration test modules, so not a big deal.

ebullient replied:
Well. I just wanted to make sure that a user would not have to do this in their app. ;)

radcortez replied:
It should have worked it `@RegisterForReflection`, but for some reason it didn't work. I'll have to look into that.
